File: ai-models/workers/api-gateway/src/routes/image.ts
import { Hono } from "hono";
import { HonoBindings } from "../env";
import { callInferenceServer } from "../lib/inferenceClient";
import { makeCacheKey } from "../lib/cache";

const image = new Hono<HonoBindings>();

const VALID_COLORS = ["red", "blue", "green", "yellow", "purple", "black"];
const VALID_SHAPES = ["circle", "square", "triangle"];

image.post("/generate", async (c) => {
  const body = await c.req.json<{
    prompt_type: "condition" | "text";
    color?: string;
    shape?: string;
    prompt?: string;
  }>();

  if (body.prompt_type === "condition") {
    if (!body.color || !VALID_COLORS.includes(body.color)) {
      return c.json({ error: "invalid_request", detail: "invalid color" }, 400);
    }
    if (!body.shape || !VALID_SHAPES.includes(body.shape)) {
      return c.json({ error: "invalid_request", detail: "invalid shape" }, 400);
    }
  } else if (body.prompt_type === "text") {
    // 단계 3(텍스트 프롬프트 기반)용 예약 분기. 현재 추론 서버는 미지원일 수 있음.
    if (!body.prompt) {
      return c.json({ error: "invalid_request", detail: "prompt is required" }, 400);
    }
  } else {
    return c.json({ error: "invalid_request", detail: "invalid prompt_type" }, 400);
  }

  const cacheKey = await makeCacheKey("/v1/image/generate", body);
  const cached = await c.env.CACHE_KV.get(cacheKey, "json");
  if (cached) {
    return c.json(cached);
  }

  const result = await callInferenceServer(c.env, "/v1/image/generate", body);
  if (!result.ok) {
    return c.json(result.data ?? { error: "inference_error" }, result.status as any);
  }

  await c.env.CACHE_KV.put(cacheKey, JSON.stringify(result.data), {
    expirationTtl: 3600,
  });

  return c.json(result.data);
});

export default image;

File: ai-models/inference-server/routers/image.py
from typing import Literal, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from auth import verify_internal_request
from config import NANO_TECH_ARTIST_VERSION
from models import nano_tech_artist

router = APIRouter(dependencies=[Depends(verify_internal_request)])

VALID_COLORS = {"red", "blue", "green", "yellow", "purple", "black"}
VALID_SHAPES = {"circle", "square", "triangle"}


class ImageGenerateRequest(BaseModel):
    prompt_type: Literal["condition", "text"]
    color: Optional[str] = None
    shape: Optional[str] = None
    prompt: Optional[str] = None


class ImageGenerateResponse(BaseModel):
    model: str
    image_base64: str
    format: str = "png"


@router.post("/generate", response_model=ImageGenerateResponse)
def generate(req: ImageGenerateRequest) -> ImageGenerateResponse:
    if req.prompt_type == "condition":
        if req.color not in VALID_COLORS or req.shape not in VALID_SHAPES:
            raise HTTPException(status_code=400, detail="invalid color or shape")
        image_b64 = nano_tech_artist.generate_image(color=req.color, shape=req.shape)
    else:
        # 단계 3(텍스트 프롬프트 기반)은 아직 미구현.
        raise HTTPException(status_code=501, detail="text-to-image not yet implemented (roadmap milestone 3)")

    return ImageGenerateResponse(
        model=f"nano-tech-artist-{NANO_TECH_ARTIST_VERSION}",
        image_base64=image_b64,
    )
